cartographer_ros/cartographer_ros/msg_conversion.cc

```diff
@@ -32,13 +32,44 @@
 #include "geometry_msgs/Vector3.h"
 #include "glog/logging.h"
 #include "nav_msgs/OccupancyGrid.h"
+#include "pcl/point_cloud.h"
+#include "pcl/point_types.h"
+#include "pcl_conversions/pcl_conversions.h"
 #include "ros/ros.h"
 #include "ros/serialization.h"
 #include "sensor_msgs/Imu.h"
 #include "sensor_msgs/LaserScan.h"
 #include "sensor_msgs/MultiEchoLaserScan.h"
 #include "sensor_msgs/PointCloud2.h"
 
+namespace {
+
+// Sizes of PCL point types have to be 4n floats for alignment, as described in
+// http://pointclouds.org/documentation/tutorials/adding_custom_ptype.php
+struct PointXYZT {
+  float x;
+  float y;
+  float z;
+  float time;
+};
+
+struct PointXYZIT {
+  PCL_ADD_POINT4D;
+  float intensity;
+  float time;
+  float unused_padding[2];
+};
+
+}  // namespace
+
+POINT_CLOUD_REGISTER_POINT_STRUCT(
+    PointXYZT, (float, x, x)(float, y, y)(float, z, z)(float, time, time))
+
+POINT_CLOUD_REGISTER_POINT_STRUCT(
+    PointXYZIT,
+    (float, x, x)(float, y, y)(float, z, z)(float, intensity,
+                                            intensity)(float, time, time))
+
 namespace cartographer_ros {
 namespace {
 
```

Comment:
By the way, I don't think this is true:

https://github.com/googlecartographer/cartographer_ros/blob/82b5eb96884b48f4e77672fdd8d169bac9882d26/cartographer_ros/cartographer_ros/msg_conversion.cc#L76-L79

If you play the bags I provided and visualize the lidar topics, rViz will happily show the point clouds, even though the last float is the [time offset](https://gist.github.com/ojura/ca9d0e50bc9cd68d33fc2316dfb4e1f8#file-timed_pointcloud2-cpp-L41), not [`1.0f`](https://github.com/googlecartographer/cartographer_ros/blob/82b5eb96884b48f4e77672fdd8d169bac9882d26/cartographer_ros/cartographer_ros/msg_conversion.cc#L198).